csharp/src/Ice/INetworkProxy.cs

```diff
@@ -98,70 +69,55 @@ public void BeginWrite(EndPoint endpoint, IList<ArraySegment<byte>> buffer)
             MemoryMarshal.Write(data.AsSpan(2, 2), ref port); // Port
             Buffer.BlockCopy(addr.Address.GetAddressBytes(), 0, data, 4, 4); // IPv4 address
             data[8] = 0x00; // User ID.
-            buffer.Add(data);
-        }
-
-        public int EndWrite(IList<ArraySegment<byte>> buffer, int bytesTransferred) =>
-            // Once the request is sent, read the response
-            bytesTransferred < buffer.GetByteCount() ? SocketOperation.Write : SocketOperation.Read;
-
-        // Read the SOCKS4 response whose size is 8 bytes.
-        public ArraySegment<byte> BeginRead() => new ArraySegment<byte>(new byte[8]);
 
-        public int EndRead(ref ArraySegment<byte> buffer, int offset) =>
-            // We're done once we read the response
-            offset < buffer.Count ? SocketOperation.Read : SocketOperation.None;
+            // Send the request.
+            await socket.SendAsync(data, SocketFlags.None, cancel);
 
-        public void Finish(ArraySegment<byte> buffer)
-        {
-            if (buffer[0] != 0x00 || buffer[1] != 0x5a)
+            // Now wait for the response.
+            await socket.ReceiveAsync(new ArraySegment<byte>(data, 0, 8), SocketFlags.None, cancel);
```

Comment:
I don't think so because the response contains 8 bytes and we have to read all of them. I've back the comment to explain that the response is 8 bytes.

Reply:
I see, we can still use `AsMemory` seems more idiomatic than `new ArraySegment`